### PR TITLE
Adds a FeatureModelBuilder and iterates depth first

### PR DESCRIPTION
**Pull request by s9latimm**
- [x] FeatureModelBuilder
- [x] depth first iterator
- [x] comments
- [x] update bindings
- [x] clang-tidy

### CODE REVIEW

File: include/vara/Feature/Feature.h [hunk 1]
```diff
@@ -1,13 +1,15 @@
...
-#include "vara/Feature/Location.h"
-#include "vara/Feature/Relationship.h"
+#include "vara/Feature/FeatureSourceRange.h"
 
-#include "llvm/ADT/SmallVector.h"
+#include "llvm/ADT/DenseSet.h"
 #include "llvm/ADT/iterator_range.h"
 #include "llvm/Support/raw_ostream.h"
 
+#include <set>
```
Comment (vulder) [inline include/vara/Feature/Feature.h:10]: seems like this is not used

File: include/vara/Feature/Feature.h [hunk 2]
```diff
@@ -20,28 +22,37 @@ namespace vara::feature {
...
+  void addAlternative(Feature *F) { Alternatives.insert(F); }
+
+  void addImplication(Feature *F) { Implications.insert(F); }
 
 protected:
-  Feature(string Name, bool Opt, std::optional<Location> Loc)
-      : Name(std::move(Name)), Opt(Opt), Loc(std::move(Loc)) {}
+  Feature(string Name, bool Opt, std::optional<FeatureSourceRange> Loc,
+          Feature *Parent)
+      : Name(std::move(Name)), Opt(Opt), Loc(std::move(Loc)), Parent(Parent) {}
```
Comment (vulder) [inline include/vara/Feature/Feature.h]: Protected section should come after public section

File: include/vara/Feature/Feature.h [hunk 3]
```diff
@@ -20,28 +22,37 @@ namespace vara::feature {
...
+
+  void addChild(Feature *F) { Children.insert(F); }
+
+  void setParent(Feature *F) { Parent = F; }
+
+  void addExclude(Feature *F) { Excludes.insert(F); }
+
+  void addAlternative(Feature *F) { Alternatives.insert(F); }
+
+  void addImplication(Feature *F) { Implications.insert(F); }
```
Comment (vulder) [inline include/vara/Feature/Feature.h]: Private sections should come after public/protected sections

File: include/vara/Feature/Feature.h [hunk 4]
```diff
@@ -52,7 +63,11 @@ class Feature {
 
   [[nodiscard]] bool isOptional() const { return Opt; }
 
-  [[nodiscard]] bool isRoot() const { return Parents.empty(); }
+  [[nodiscard]] virtual bool isBinary() const = 0;
+
+  [[nodiscard]] virtual bool isNumeric() const = 0;
```
Comment (vulder) [inline include/vara/Feature/Feature.h]: Why to we implement specific check functions when we use inheritance? If we need this information we should impl llvm style type info with kind enums. Then we can write `llvm::isa<BinaryFeature *>()`

File: include/vara/Feature/Feature.h [hunk 5]
```diff
@@ -203,30 +169,41 @@ class Feature {
...
-           llvm::StringRef(F.getName()).lower();
+  bool operator==(const vara::feature::Feature &F) const {
+    return getName().lower() == F.getName().lower();
   }
 
-  bool operator<(const Feature &F) const {
-    return llvm::StringRef(this->getName()).lower() <
-           llvm::StringRef(F.getName()).lower();
+  struct FeatureDepthFirstComparator {
+    bool operator()(vara::feature::Feature *, vara::feature::Feature *) const;
```
Comment (vulder) [inline include/vara/Feature/Feature.h]: Is a ptr here required? I would assume `const &` should be enough here

File: include/vara/Feature/FeatureModel.h [hunk 6]
```diff
@@ -17,7 +19,8 @@ namespace vara::feature {
 class FeatureModel {
 public:
   using FeatureMapTy = llvm::StringMap<std::unique_ptr<Feature>>;
-  using ConstraintsTy = std::vector<std::vector<std::pair<Feature *, bool>>>;
+  using ConstraintTy = llvm::SmallVector<std::pair<std::string, bool>, 3>;
+  using ConstraintsTy = std::vector<ConstraintTy>;
 
 private:
```
Comment (vulder) [inline include/vara/Feature/FeatureModel.h]: same public/protected/private order problem

File: include/vara/Feature/FeatureModel.h [hunk 7]
```diff
@@ -68,29 +69,135 @@ class FeatureModel {
     }
   };
 
-  FeatureModelIter begin() { return FeatureModelIter(Features.begin()); }
+public:
+  using feature_model_iterator = FeatureModelIter;
+  using const_feature_model_iterator = FeatureModelIter;
+
+  feature_model_iterator begin() { return FeatureModelIter(Features.begin()); }
```
Comment (vulder) [inline include/vara/Feature/FeatureModel.h]: Why is the FeatureModelIter private but exposed throught the public interface?

File: include/vara/Feature/FeatureModel.h [hunk 8]
```diff
@@ -68,29 +69,135 @@ class FeatureModel {
...
+
+  [[nodiscard]] const_feature_model_iterator begin() const {
+    return FeatureModelIter(Features.begin());
+  }
 
   FeatureModelIter end() { return FeatureModelIter(Features.end()); }
 
+  [[nodiscard]] const_feature_model_iterator end() const {
+    return FeatureModelIter(Features.end());
+  }
```
Comment (vulder) [inline include/vara/Feature/FeatureModel.h:81]: Please add also add llvm style range function `features`

File: include/vara/Feature/FeatureModel.h [hunk 9]
```diff
@@ -68,29 +69,135 @@ class FeatureModel {
...
-    for (const auto &F : this->Features) {
-      F.second->dump();
-      llvm::outs() << "\n";
+  void dump() const;
+
+  class FeatureModelBuilder {
+    using EdgeMapType =
+        typename llvm::StringMap<llvm::SmallVector<std::string, 3>>;
+
+    std::string VmName;
```
Comment (vulder) [inline include/vara/Feature/FeatureModel.h]: private/public order

File: include/vara/Feature/FeatureModel.h [hunk 10]
```diff
@@ -68,29 +69,135 @@ class FeatureModel {
...
+          .try_emplace(
+              Key, std::make_unique<BinaryFeature>(Key, Opt, std::move(Loc)))
+          .second;
+    }
+
+    FeatureModelBuilder *addChild(const std::string &P, const std::string &C) {
+      Children[P].push_back(C);
+      Parents[C] = P;
+      return this;
+    }
```
Comment (vulder) [inline include/vara/Feature/FeatureModel.h]: Does it make sense to also add overloads for when we have a `Feature &` and what to add a child etc.?

File: include/vara/Feature/FeatureModelParser.h [hunk 11]
```diff
@@ -99,28 +99,15 @@ class FeatureModelXmlParser : public FeatureModelParser {
 
 private:
```
Comment (vulder) [inline include/vara/Feature/FeatureModelParser.h]: This private section should also be at the bottom

File: include/vara/Feature/FeatureSort.h [hunk 12]
```diff
@@ -0,0 +1,40 @@
+#ifndef VARA_FEATURE_FEATURESORT_H
+#define VARA_FEATURE_FEATURESORT_H
+
+#include "FeatureModel.h"
+
+namespace vara::feature {
+
+class FeatureSort {
```
Comment (vulder) [inline include/vara/Feature/FeatureSort.h]: Sounds like an algorithm. I would prefer name like `OrderedFeatureVector`

File: include/vara/Feature/FeatureSort.h [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+public:
+  using feature_heap_iterator = typename std::vector<Feature *>::iterator;
+  using const_feature_heap_iterator =
+      typename std::vector<Feature *>::const_iterator;
+
+private:
```
Comment (vulder) [inline include/vara/Feature/FeatureSort.h]: private to the bottom

File: include/vara/Feature/FeatureSort.h [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+  std::vector<Feature *> Features;
+
+public:
+  template <typename T> FeatureSort(T FM) {
+    for (auto F = FM->begin(); F != FM->end(); ++F) {
+      Features.push_back(*F);
+    }
+    std::sort(Features.begin(), Features.end(),
+              Feature::FeatureDepthFirstComparator());
+  }
```
Comment (vulder) [inline include/vara/Feature/FeatureSort.h]: Currently, this is not a heap. Didn't you want to make it a heap?

File: include/vara/Feature/FeatureSort.h (hunk 12 above)
Comment (vulder) [inline include/vara/Feature/FeatureSort.h]: Currently, there are insertion methods missing. And I also do not see how this is used in the FM. I'll let you finish the impl before I'll continue to look at stuff :)

**Review by vulder — CHANGES_REQUESTED** [review]
sry, for the public/protected/private stuff I missed this last time

File: include/vara/Feature/FeatureSort.h (hunk 14 above)
Comment (s9latimm) [reply include/vara/Feature/FeatureSort.h]: Building a heap I would need to call `sort_heap` to receive a sorted collection while losing the heap property. As a custom iterator does not seem reasonable at this point, I used a  comparator with insertion at `std::upper_bound` instead.

File: bindings/python/tests/feature/test_feature.py [hunk 15]
```diff
@@ -28,74 +29,18 @@ def test_is_root(self):
...
-    def test_add_alternative(self):
-        """ Checks if a feature is correctly added as an alternative from
-        another. """
-        root_feature = feature.BinaryFeature("root", False)
-        test_feature = feature.BinaryFeature("Test", False)
-
-        test_feature.add_alternative(root_feature)
-
-        self.assertTrue(test_feature.is_alternative(root_feature))
+    # def test_iter_children(self):
```
Comment (vulder) [inline bindings/python/tests/feature/test_feature.py]: why is this case outcommented?

File: bindings/python/vara-feature/pybind_FeatureSourceRange.cpp [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+namespace fs = std::experimental::filesystem;
+#else
+#include <filesystem>
+namespace fs = std::filesystem;
+#endif
+
+namespace vf = vara::feature;
+namespace py = pybind11;
+
+void init_feature_module_location(py::module &M) {
```
Comment (vulder) [inline bindings/python/vara-feature/pybind_FeatureSourceRange.cpp]: `init_feature_location_module`

File: bindings/python/vara-feature/pybind_FeatureSourceRange.cpp [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+      .def_property_readonly(
+          "path",
+          [](vf::FeatureSourceRange &Loc) { return Loc.getPath().string(); },
+          R"pbdoc(Path to the source file)pbdoc")
+      .def("get_start", &vf::FeatureSourceRange::getStart,
+           py::return_value_policy::reference,
+           R"pbdoc(Get the start `LineColumnOffset` of this `Location`.)pbdoc")
+      .def("get_end", &vf::FeatureSourceRange::getEnd,
+           py::return_value_policy::reference,
+           R"pbdoc(Get the end `LineColumnOffset` of this `Location`.)pbdoc")
```
Comment (vulder) [inline bindings/python/vara-feature/pybind_FeatureSourceRange.cpp]: If I remember correctly pascal migrated accessors to python properties, we should use this here too.

File: include/vara/Feature/Feature.h [hunk 18]
```diff
@@ -21,44 +22,25 @@ namespace vara::feature {
...
-  FeatureListType Children;
-  FeatureListType Excludes;
-  FeatureListType Implications;
-  FeatureListType Alternatives;
-  RelationshipListTy Relationships;
-
-protected:
-  Feature(string Name, bool Opt, std::optional<Location> Loc)
-      : Name(std::move(Name)), Opt(Opt), Loc(std::move(Loc)) {}
+  enum FeatureType { BINARY, NUMERIC };
```
Comment (vulder) [inline include/vara/Feature/Feature.h]:
```suggestion
  enum FeatureKind { FK_Binary, FK_Numeric };
```

File: include/vara/Feature/Feature.h [hunk 19]
```diff
@@ -21,44 +22,25 @@ namespace vara::feature {
...
 
-public:
   Feature(const Feature &) = delete;
-  Feature &operator=(const Feature &) = delete;
   virtual ~Feature() = default;
 
+  [[nodiscard]] FeatureType getType() const { return T; }
```
Comment (vulder) [inline include/vara/Feature/Feature.h]:
```suggestion
  [[nodiscard]] FeatureKind getKind() const { return T; }
```

File: include/vara/Feature/Feature.h [hunk 20]
```diff
@@ -203,34 +136,90 @@ class Feature {
   }
   [[nodiscard]] const_feature_iterator end() const { return children_end(); }
 
-  [[nodiscard]] Location* getLocation() {
+  //===--------------------------------------------------------------------===//
+  // Operators
+  Feature &operator=(const Feature &) = delete;
```
Comment (vulder) [inline include/vara/Feature/Feature.h]: Copy assignment operator should be moved to the ctor if possible.

File: include/vara/Feature/Feature.h [hunk 21]
```diff
@@ -203,34 +136,90 @@ class Feature {
...
 
-  bool operator<(const Feature &F) const {
-    return llvm::StringRef(this->getName()).lower() <
-           llvm::StringRef(F.getName()).lower();
-  }
+  LLVM_DUMP_METHOD
+  void dump() const { llvm::outs() << toString() << "\n"; }
+
+protected:
+  Feature(FeatureType T, string Name, bool Opt,
```
Comment (vulder) [inline include/vara/Feature/Feature.h]:
```suggestion
  Feature(FeatureKind Kind, string Name, bool Opt,
```

File: include/vara/Feature/Feature.h [hunk 22]
```diff
@@ -203,34 +136,90 @@ class Feature {
...
+
+  void addChild(Feature *F) { Children.insert(F); }
+
+  void setParent(Feature *F) { Parent = F; }
+
+  void addExclude(Feature *F) { Excludes.insert(F); }
+
+  void addAlternative(Feature *F) { Alternatives.insert(F); }
+
+  void addImplication(Feature *F) { Implications.insert(F); }
```
Comment (vulder) [inline include/vara/Feature/Feature.h]:
```suggestion
  void addChild(Feature *F) { Children.insert(F); }

  void setParent(Feature *F) { Parent = F; }

  void addExclude(Feature *F) { Excludes.insert(F); }

  void addAlternative(Feature *F) { Alternatives.insert(F); }

  void addImplication(Feature *F) { Implications.insert(F); }
  
  FeatureType T;
  string Name;
  bool Opt;
  std::optional<FeatureSourceRange> Loc;
  Feature *Parent;
  FeatureSetType Children;
  FeatureSetType Excludes;
  FeatureSetType Implications;
  FeatureSetType Alternatives;
```
Nit: I would prefer to have all members at the end of the class.

File: include/vara/Feature/Feature.h [hunk 23]
```diff
@@ -203,34 +136,90 @@ class Feature {
...
-                std::optional<Location> Loc = std::nullopt)
-      : Feature(std::move(Name), Opt, std::move(Loc)) {}
+  BinaryFeature(string Name, bool Opt = false,
+                std::optional<FeatureSourceRange> Loc = std::nullopt,
+                Feature *Parent = nullptr, FeatureSetType Children = {},
+                FeatureSetType Excludes = {}, FeatureSetType Implications = {},
+                FeatureSetType Alternatives = {})
+      : Feature(BINARY, std::move(Name), Opt, std::move(Loc), Parent,
+                std::move(Children), std::move(Excludes),
+                std::move(Implications), std::move(Alternatives)) {}
```
Comment (vulder) [inline include/vara/Feature/Feature.h]: Does it make sense to make the ctor protected and only offer a factory function on the `FeatureModel`? Otherwise, we cannot guarantee that the values set in the ctor are consistent with the rest of the model.

File: include/vara/Feature/FeatureModel.h [hunk 24]
```diff
@@ -1,37 +1,40 @@
...
-      : Name(std::move(Name)), RootPath(std::move(RootPath)),
-        Features(std::move(Features)), Constraints(std::move(Constraints)),
-        Root(this->Features["root"].get()) {}
+               Feature *Root)
+      : Name(std::move(Name)), RootPath(std::move(RootPath)), Root(Root),
+        Features(std::move(Features)) {
+    // Insert all values into ordered data structure.
+    for (const auto &KV : this->Features) {
+      OrderedFeatures.insert(KV.getValue().get());
+    }
```
Comment (vulder) [inline include/vara/Feature/FeatureModel.h:34]: Update to new insert method. (see below)

File: include/vara/Feature/FeatureSourceRange.h [hunk 25]
```diff
@@ -0,0 +1,102 @@
...
+//===----------------------------------------------------------------------===//
+
+class FeatureSourceRange {
+public:
+  class FeatureSourceLocation {
+
+  public:
+    FeatureSourceLocation(int Line, int Column) : Line(Line), Column(Column) {}
+
+    void setLineNumber(int L) { this->Line = L; }
```
Comment (vulder) [inline include/vara/Feature/FeatureSourceRange.h]:
```suggestion
    void setLineNumber(int LineNumber) { this->Line = LineNumber; }
```

File: include/vara/Feature/FeatureSourceRange.h [hunk 26]
```diff
@@ -0,0 +1,102 @@
...
+    [[nodiscard]] int getLineNumber() const { return this->Line; }
+
+    void setColumnOffset(int C) { this->Column = C; }
```
Comment (vulder) [inline include/vara/Feature/FeatureSourceRange.h]:
```suggestion
    void setColumnOffset(int ColumnOffset) { this->Column = ColumnOffset; }
```

File: include/vara/Feature/OrderedFeatureVector.h [hunk 27]
```diff
@@ -0,0 +1,46 @@
...
+  OrderedFeatureVector() = default;
+
+  /// Insert feature while preserving ordering.
+  void insert(Feature *F) {
+    Features.insert(
+        std::upper_bound(Features.begin(), Features.end(), F,
+                         [](vara::feature::Feature *A,
+                            vara::feature::Feature *B) { return *A < *B; }),
+        F);
+  }
```
Comment (vulder) [inline include/vara/Feature/OrderedFeatureVector.h:34]: It would be helpful to provide different overloads for insert, especially, the one that takes and iterator range :wink: .

File: lib/Feature/Feature.cpp [hunk 28]
```diff
@@ -1,57 +1,34 @@
 #include "vara/Feature/Feature.h"
 
+#include <experimental/iterator>
```
Comment (vulder) [inline lib/Feature/Feature.cpp]:
```suggestion
#include <iterator>
```
Why would we need experimental stuff here?

File: unittests/Feature/Feature.cpp [hunk 29]
```diff
@@ -1,59 +1,79 @@
...
+  auto FM = FeatureModelBuilder().buildSimpleFeatureModel(
+      {{"a", "aa"}, {"root", "b"}, {"root", "a"}});
+  assert(FM);
+
+  EXPECT_LT(*FM->getFeature("root"), *FM->getFeature("a"));
+  EXPECT_LT(*FM->getFeature("root"), *FM->getFeature("aa"));
+  EXPECT_LT(*FM->getFeature("root"), *FM->getFeature("b"));
+  EXPECT_LT(*FM->getFeature("a"), *FM->getFeature("aa"));
+  EXPECT_LT(*FM->getFeature("a"), *FM->getFeature("b"));
+  EXPECT_LT(*FM->getFeature("aa"), *FM->getFeature("b"));
```
Comment (vulder) [inline unittests/Feature/Feature.cpp]: The less than impl tries to implement a quite complicated relationship, could you please split this case up into different test cases that cover the different scenarios.

File: lib/Feature/FeatureModel.cpp [hunk 30]
```diff
@@ -0,0 +1,148 @@
...
+}
+
+void FeatureModelBuilder::buildConstraints() {
+  for (const auto &F : Features.keys()) {
+    for (const auto &E : Excludes[F]) {
+      Features[F]->addExclude(Features[E].get());
+    }
+  }
+
+  for (const auto &C : Constraints) {
```
Comment (vulder) [inline lib/Feature/FeatureModel.cpp]: If we iterate over lists here we should use a better name than `C`. What is the actual type of `C`? maybe it could help to spell it out instead of auto.

File: lib/Feature/FeatureModel.cpp [hunk 31]
```diff
@@ -0,0 +1,148 @@
...
+      if (B) {
+        for (const auto &P : C) {
+          for (const auto &PP : C) {
+            if (P.first != PP.first) {
+              Features[P.first]->addAlternative(Features[PP.first].get());
+            }
+          }
+        }
+      }
+    }
```
Comment (vulder) [inline lib/Feature/FeatureModel.cpp:86]: What happens in the else case? If `C.size()` is `0` or `1`?

File: lib/Feature/Feature.cpp [hunk 32]
```diff
@@ -78,4 +52,45 @@ std::string NumericFeature::toString() const {
 
 std::string BinaryFeature::toString() const { return Feature::toString(); }
 
+bool Feature::operator<(const vara::feature::Feature &F) const {
```
Comment (vulder) [inline lib/Feature/Feature.cpp]: For a `operator<` this is quite complex, so we should test this well, do the test cover all edge cases?

**Review by vulder — CHANGES_REQUESTED** [review]
Found a few minor things

File: include/vara/Feature/FeatureModel.h (hunk 10 above)
Comment (s9latimm) [reply include/vara/Feature/FeatureModel.h]: This method / builder is supposed to be only used while parsing / iteratively building a FM, so it makes not that much sense to build those connections for an already existing `Feature` by hand or even hold such an object in the first place. I added `addFeature(Feature&)` instead which basically behaves like inserting all related connections one after another, which can be used to duplicate existing subtrees between different FMs.

File: include/vara/Feature/Feature.h (hunk 23 above)
Comment (s9latimm) [reply include/vara/Feature/Feature.h]: FM consistency is assured through that Features will only be added with `FeatureModel::addFeature(std::unique_ptr<Feature>) : bool` checking whether a Feature can be inserted without violating any rules (currently only rudimentary). On the other hand it may be handy to create (unrelated) features beforehand to store information (not only for testing), building the actual FM afterwards (as the builder provides corresponding methods).

File: lib/Feature/Feature.cpp (hunk 32 above)
Comment (s9latimm) [reply lib/Feature/Feature.cpp]: I tested all pairwise relations on binary trees with height 2, basic cases and also unrelated features.

File: bindings/python/vara-feature/pybind_Feature.cpp [hunk 33]
```diff
@@ -55,42 +47,37 @@ void init_feature_module_feature(py::module &M) {
...
-            return py::make_iterator(F.excludes_begin(), F.excludes_end());
+            return py::make_iterator(F.implications_begin(),
+                                     F.implications_end());
           },
           py::keep_alive<0, 1>(), R"pbdoc(Implicated features)pbdoc")
-      .def(
-          "add_implication", &vf::Feature::addImplication,
-          R"pbdoc(Implicate another Feature, i.e., selecting this feature implies the otherone.)pbdoc")
-      .def("implicates", &vf::Feature::implies,
+      .def("is_implied", &vf::Feature::isImplied,
```
Comment (vulder) [inline bindings/python/vara-feature/pybind_Feature.cpp]:
```suggestion
      .def("is_implied_by", &vf::Feature::isImplied,
```
This slight name change should make it a bit clearer

File: unittests/Feature/OrderedFeatureVector.cpp [hunk 34]
```diff
@@ -0,0 +1,41 @@
...
+
+TEST(OrderedFeatureVector, insertVariadic) {
+  auto FM = FeatureModelBuilder().buildSimpleFeatureModel(
+      {{"root", "a"}, {"root", "b"}});
+  OrderedFeatureVector OFV;
+
+  OFV.insert(FM->getFeature("a"), FM->getFeature("root"), FM->getFeature("b"));
+
+  EXPECT_EQ(OFV.size(), FM->size());
+  EXPECT_EQ(*OFV.begin(), FM->getFeature("root"));
```
Comment (vulder) [inline unittests/Feature/OrderedFeatureVector.cpp]: All tests only check if the first one is root but do not verify if the others are in correct order.

File: include/vara/Feature/FeatureModel.h [hunk 35]
```diff
@@ -1,96 +1,206 @@
...
+  /// Try to add \a NumericFeature.
+  bool addFeature(const std::string &Key,
+                  const NumericFeature::ValuesVariantType &Values,
+                  bool Opt = false,
+                  std::optional<FeatureSourceRange> Loc = std::nullopt) {
+    return Features
+        .try_emplace(Key, std::make_unique<NumericFeature>(Key, Values, Opt,
+                                                           std::move(Loc)))
+        .second;
+  }
```
Comment (vulder) [inline include/vara/Feature/FeatureModel.h]: We should merge these functions into a factory method and rename them to `make_feature` or `create_feature`, to not be confused with the `addFeature(Feature &)` method, which has a different behavior (Add a feature by copying it vs creating a new feature). Furthermore, `Key` is not a really good name here, it is the key for the map but from the API users side it is the feature name, and we should not confuse the user by naming things based on our implementation details.

Regarding the factory method:
```
    /// Try to create and add a new \a Feature to the \a FeatureModel.
  ///
  /// \param[in] FeatureName name of the \a Feature
  /// \param[in] FurtherArgs further arguments that should be passed to the
  ///                        \a Feature constructor
  ///
  /// \returns true, if the feature could be inserted into the \a FeatureModel
  template <
      typename FeatureTy, typename... Args,
      typename std::enable_if_t<std::is_base_of_v<Feature, FeatureTy>, int>>
  bool makeFeature(const std::string &FeatureName, Args... FurtherArgs) {
    return Features
        .try_emplace(FeatureName, std::make_unique<FeatureTy>(
                                      FeatureName, std::forward(FurtherArgs)...))
        .second;
  }
```
This allows users to create any kind of feature as long as it is derived from `Feature` and the passed args match one of the specific Features ctors.

File: include/vara/Feature/FeatureModel.h [hunk 36]
```diff
@@ -1,96 +1,206 @@
...
+    Children[ParentKey].insert(Key);
+    Parents[Key] = ParentKey;
+    return this;
+  }
+
+  FeatureModelBuilder *addExclude(const std::string &Key,
+                                  const std::string &ExcludeKey) {
+    Excludes[Key].insert(ExcludeKey);
+    return this;
+  }
```
Comment (vulder) [inline include/vara/Feature/FeatureModel.h]: What happens in these functions if the `Key` is not a Feature name that was inserted before?

PS: `Key` is also a bad name for this param, we should rename this.

File: include/vara/Feature/OrderedFeatureVector.h [hunk 37]
```diff
@@ -0,0 +1,51 @@
...
+class OrderedFeatureVector {
+public:
+  using ordered_feature_iterator = typename std::vector<Feature *>::iterator;
+  using const_ordered_feature_iterator =
+      typename std::vector<Feature *>::const_iterator;
+
+  OrderedFeatureVector() = default;
+
+  /// Insert feature while preserving ordering.
+  void insert(Feature *F);
```
Comment (vulder) [inline include/vara/Feature/OrderedFeatureVector.h:23]:
```suggestion
  void insert(Feature *F);
  
```

File: include/vara/Feature/OrderedFeatureVector.h [hunk 38]
```diff
@@ -0,0 +1,51 @@
...
+  template <typename... Args> void insert(Feature *F, Args... FF) {
+    insert(F);
+    insert(FF...);
+  }
+  template <typename T> void insert(llvm::iterator_range<T> FF) {
```
Comment (vulder) [inline include/vara/Feature/OrderedFeatureVector.h:30]:
```suggestion

  template <typename T> void insert(llvm::iterator_range<T> FF) {
```

File: include/vara/Feature/Feature.h [hunk 39]
```diff
@@ -21,44 +23,26 @@ namespace vara::feature {
...
-  FeatureListType Children;
-  FeatureListType Excludes;
-  FeatureListType Implications;
-  FeatureListType Alternatives;
-  RelationshipListTy Relationships;
-
-protected:
-  Feature(string Name, bool Opt, std::optional<Location> Loc)
-      : Name(std::move(Name)), Opt(Opt), Loc(std::move(Loc)) {}
+  enum FeatureKind { FK_BINARY, FK_NUMERIC };
```
Comment (vulder) [inline include/vara/Feature/Feature.h]: Maybe it would be better to use a enum class here so integer constant can not be confused with the feature kinds.

**Review by vulder — CHANGES_REQUESTED** [review]
Test coverage looks much better now :smiley:

File: include/vara/Feature/FeatureModel.h (hunk 36 above)
Comment (s9latimm) [reply include/vara/Feature/FeatureModel.h]: Doing so is totally fine, as the consistency (existence) is asserted when building the final FM. As the parser can not assure that features are only referenced if they are already processed (like mutual exclusion), this can not be asserted in this method either.

**Review by padupr — APPROVED** [review]

**Review by vulder — APPROVED** [review]
lgtm
